includes/sanitizers/class-amp-style-sanitizer.php

```diff
@@ -349,230 +431,741 @@ private function process_link_element( DOMElement $element ) {
+		if ( empty( $src_properties ) ) {
+			return;
+		}
 
-		// Normalize whitespace and filter rules.
-		foreach ( $styles as $index => $rule ) {
-			$tuple = preg_split( '/\s*:\s*/', $rule, 2 );
-			if ( 2 !== count( $tuple ) ) {
+		foreach ( $src_properties as $src_property ) {
+			$value = $src_property->getValue();
+			if ( ! ( $value instanceof RuleValueList ) ) {
 				continue;
 			}
 
-			list( $property, $value ) = $this->filter_style( $tuple[0], $tuple[1] );
-			if ( empty( $property ) || empty( $value ) ) {
+			/*
+			 * The CSS Parser parses a src such as:
+			 *
+			 *    url(data:application/font-woff;...) format('woff'),
+			 *    url('Genericons.ttf') format('truetype'),
+			 *    url('Genericons.svg#genericonsregular') format('svg')
+			 *
+			 * As a list of components consisting of:
+			 *
+			 *    URL,
+			 *    RuleValueList( CSSFunction, URL ),
+			 *    RuleValueList( CSSFunction, URL ),
+			 *    CSSFunction
+			 *
+			 * Clearly the components here are not logically grouped. So the first step is to fix the order.
+			 */
+			$sources = array();
+			foreach ( $value->getListComponents() as $component ) {
+				if ( $component instanceof RuleValueList ) {
+					$subcomponents = $component->getListComponents();
+					$subcomponent  = array_shift( $subcomponents );
+					if ( $subcomponent ) {
+						if ( empty( $sources ) ) {
+							$sources[] = array( $subcomponent );
+						} else {
+							$sources[ count( $sources ) - 1 ][] = $subcomponent;
+						}
+					}
+					foreach ( $subcomponents as $subcomponent ) {
+						$sources[] = array( $subcomponent );
+					}
+				} else {
+					if ( empty( $sources ) ) {
+						$sources[] = array( $component );
+					} else {
+						$sources[ count( $sources ) - 1 ][] = $component;
+					}
+				}
+			}
+
+			/**
+			 * Source URL lists.
+			 *
+			 * @var URL[] $source_file_urls
+			 * @var URL[] $source_data_urls
+			 */
+			$source_file_urls = array();
+			$source_data_urls = array();
+			foreach ( $sources as $i => $source ) {
+				if ( $source[0] instanceof URL ) {
+					if ( 'data:' === substr( $source[0]->getURL()->getString(), 0, 5 ) ) {
+						$source_data_urls[ $i ] = $source[0];
+					} else {
+						$source_file_urls[ $i ] = $source[0];
+					}
+				}
+			}
+
+			// Convert data: URLs into regular URLs, assuming there will be a file present (e.g. woff fonts in core themes).
+			if ( empty( $source_file_urls ) ) {
 				continue;
 			}
+			$source_file_url = current( $source_file_urls );
+			foreach ( $source_data_urls as $i => $data_url ) {
+				$mime_type = strtok( substr( $data_url->getURL()->getString(), 5 ), ';' );
+				if ( ! $mime_type ) {
+					continue;
+				}
+				$extension   = preg_replace( ':.+/(.+-)?:', '', $mime_type );
+				$guessed_url = preg_replace(
+					':(?<=\.)\w+(\?.*)?(#.*)?$:', // Match the file extension in the URL.
+					$extension,
+					$source_file_url->getURL()->getString(),
+					1,
+					$count
+				);
+				if ( $count ) {
```

Comment:
We should really confirm whether or not the guessed file does exist. While it does in the core themes for Genericons and in core itself for Dashicons, another theme may decide to not bundle the WOFF file at all.